service/cziti/dns-proxy.go

```diff
@@ -341,69 +344,117 @@ outer:
 				} else {
 					if len(respChan) == cap(respChan) {
 						log.Warn("respChan will be blocked. If this warning is continuously displayed please report")
+					} else {
+						log.Tracef("Response received, respChan len: %d, cap: %d ", len(respChan), cap(respChan))
 					}
+
 					respChan <- resp[:n]
 				}
 			}
 		}(proxy)
 	}
 
 	reqs := make(map[uint32]*proxiedReq)
+	var dnsReqMutex = &sync.Mutex{}
 
-	log.Debug("Upstream DNS proxy loop begins")
-	for {
-		select {
-		case pr := <-proxiedRequests:
-			id := (uint32(pr.req.Id) << 16) | uint32(pr.req.Question[0].Qtype)
-			reqs[id] = pr
-			b, _ := pr.req.Pack()
-			for _, proxy := range dnsUpstreams {
-				if _, err := proxy.Write(b); err != nil {
-
-					_ = proxy.Close() //first thing - close the proxy connection
-
-					// when this hits - it never seems to recover. throw a panic which will get recovered
-					// via the defer specified above and try to recreate the connections. this is a heavy
-					// handed way of reconnecting to the DNS server but it should be infrequent
-					log.Panicf("failed to proxy DNS to %s %s %v. %v",
-						dns.Type(pr.req.Question[0].Qtype),
-						pr.req.Question[0].Name,
-						proxy.RemoteAddr(),
-						err)
-				} else {
-					log.Tracef("Proxied request sent to %v from ipv%d listener", proxy.RemoteAddr(), pr.ipVer)
+	dnsReqChannel := make(chan struct{})
+	ticker := time.NewTicker(5 * time.Second)
+	defer func() {
+		close(dnsReqChannel)
+		log.Tracef("Exiting dns request handling routines.")
+	}()
+	go func() {
+	CleanUpIteration:
+		for {
+			select {
```

Comment:
i don't think you need this select/case here, can you can just stop the ticker in the defer and thus can just range over the ticker, right?

Reply:
I tested it that way. When I remove this select/case, ticket will stop ticking, but the go routine is not exiting.

Reply:
how'd you code it up? did you 'range' over the ticker.C ?  this worked for me:  https://play.golang.org/p/MO3ndiv5qR

```
    ticker := time.NewTicker(time.Millisecond * 500)
    go func() {
        for t := range ticker.C {
            fmt.Println("Tick at", t)
        }
    }()

    // Tickers can be stopped like timers. Once a ticker
    // is stopped it won't receive any more values on its
    // channel. We'll stop ours after 1600ms.
    time.Sleep(time.Millisecond * 1600)
    ticker.Stop()
    fmt.Println("Ticker stopped")
```

Reply:
I added a print statement after the for loop in this example and added a sleep for 5 seconds after the ticker.Stop() function. The print statement after the for loop never got executed. So the go routine is not exiting until the main program exits.

Reply:
Gotcha. That's a bummer, but good to know.